[HUDI-8182]Cache internalSchema for hive read, avoid each split reloa…

muyihao opened the pull request:
…d active timeline.

### Change Logs
Addressing this issue: [11723](https://github.com/apache/hudi/issues/11723)
When using MR to read Hudi, use a global InternalSchema to avoid each reader listing the metadata directory to obtain the InternalSchema, which places a significant load on the HDFS NameNode.

### Impact

Reduced metadata listing frequency to alleviate pressure on the HDFS NameNode.

### Risk level (write none, low medium or high below)

low

### Documentation Update

N/A

### Contributor's checklist

- [ ] Read through [contributor's guide](https://hudi.apache.org/contribute/how-to-contribute)
- [ ] Change Logs and Impact were stated clearly
- [ ] Adequate tests were added if applicable
- [ ] CI passed

danny0405 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/hive/HoodieCombineHiveInputFormat.java:
```diff
@@ -375,6 +389,45 @@ public InputSplit[] getSplits(JobConf job, int numSplits) throws IOException {
     Utilities.clearWorkMapForConf(job);
 
+    // build internal schema for the query
+    if (job.getBoolean(INTERNAL_SCHEMA_CACHE_ENABLE, true)) {
```
Should it always be true because this is a deterministic optimization?

muyihao replied:
yes, you are right

danny0405 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/hive/HoodieCombineHiveInputFormat.java:
```diff
@@ -375,6 +389,45 @@ public InputSplit[] getSplits(JobConf job, int numSplits) throws IOException {
+          HoodieStorage storage = null;
+          try {
+            storage = new HoodieHadoopStorage(path.getFileSystem(job));
+            Option<StoragePath> tablePath = TablePathUtils.getTablePath(storage, HadoopFSUtils.convertToStoragePath(path));
```
In which case there are multiple table paths?

muyihao replied:
> In which case there are multiple table paths?

When we perform a query on two different Hudi tables, such as union, join, or subquery, there will be multiple table paths.

danny0405 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/hive/HoodieCombineHiveInputFormat.java:
```diff
@@ -375,6 +389,45 @@ public InputSplit[] getSplits(JobConf job, int numSplits) throws IOException {
+          job.set(INTERNAL_SCHEMA_CACHE_PREFIX + "." + k, v);
+        });
+
+        job.setBoolean(INTERNAL_SCHEMA_CACHE_VISIT, true);
```
Can we just check if the cache key exists instead?

muyihao replied:
Nice suggestion. I will make the changes as recommended

danny0405 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/SchemaEvolutionContext.java:
```diff
@@ -98,21 +101,52 @@ public SchemaEvolutionContext(InputSplit split, JobConf job) throws IOException
+      }
+    } else {
+      try {
+        TableSchemaResolver schemaUtil = new TableSchemaResolver(metaClient);
```
This is invalid code path because the table itself is invalid.

yihua commented:
@muyihao any update on addressing the comments in the PR?

muyihao commented:
> @muyihao any update on addressing the comments in the PR?

Currently working on verifying whether HoodieCombineHiveInputFormat can handle multiple Hudi tables and making modifications based on the comments.

danny0405 commented:
Hi, @muyihao Is this ready for review now?

muyihao commented:
> Hi, @muyihao Is this ready for review now?

@danny0405 Yes, sorry for the late reply. Thank you for your help with the review. Currently, the modifications cache the latest schema for different tables when there are multiple table paths. However, this does not completely avoid loading the ActiveTimeline. When reading COW, it will still trigger the loading of the active timeline when searching for the fileSchema.

danny0405 commented:
> When reading COW, it will still trigger the loading of the active timeline when searching for the fileSchema.

Can you show us the line that triggers this logic?

muyihao commented:
> > When reading COW, it will still trigger the loading of the active timeline when searching for the fileSchema.
> 
> Can you show us the line that triggers this logic?

<img width="1087" alt="image" src="https://github.com/user-attachments/assets/d64428ab-89b3-4b2e-bcea-ede3e2e7c9b3">

`InternalSchemaCache#searchSchemaAndCache` will trigger `InternalSchemaCache#getSchemaByReadingCommitFile` or `InternalSchemaCache##getHistoricalSchemas`, both may load active timeline

danny0405 commented:
@muyihao Thanks for pinning it out, the `metaClient.getActiveTimeline` also choose to return the cached timeline first, for one metaClient instantce, the list would happen only once, so I guess it is not a problem if the metaClient can be shared for multiple input splits.

BTW, while I reviewing the code, I found a bug for `FileBasedInternalSchemaStorageManager`, the passed in metaClient does not always work if the base path of it is not end up with `.schema`, we should fix it like this:

```java
  public FileBasedInternalSchemaStorageManager(HoodieTableMetaClient metaClient) {
    this.baseSchemaPath = new StoragePath(metaClient.getMetaPath(), SCHEMA_NAME);
    this.storage = metaClient.getStorage();
    this.metaClient = metaClient.getBasePath().getName().equalsIgnoreCase(SCHEMA_NAME) ? metaClient : null;
  }
```

danny0405 commented:
@muyihao Thanks, I have applied a patch again the master code(instead of your patch) to address the timeline list issue, you can merge the patch first then your Job conf set up logic, then it should be working:
[8182.patch.zip](https://github.com/user-attachments/files/17082708/8182.patch.zip)

muyihao commented:
@danny0405 Thanks for the help, I have merged it.

danny0405 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/hive/HoodieCombineHiveInputFormat.java:
```diff
@@ -375,6 +388,38 @@ public InputSplit[] getSplits(JobConf job, int numSplits) throws IOException {
+          if (schema.isPresent()) {
+            LOG.info("Set internal schema and avro schema of path: " + path.toString());
+            job.set(INTERNAL_SCHEMA_CACHE_KEY_PREFIX + "." + path, SerDeHelper.toJson(schema.get()));
+            job.set(SCHEMA_CACHE_KEY_PREFIX + "." + path, schemaUtil.getTableAvroSchema().toString());
```
We cache the schema by path or by table name? The table name looks more straight-forward. And the schemaUtil.getTableAvroSchema() should be invoked first to set up the commit metadata cache to avoid redundant commit metadata deserialization.

muyihao replied:
@danny0405 That's a great suggestion. When we handle tables from two different databases together, there might be the same table name. For example, dev.table_xx and online.table_xx, where table_xx is the same. Should we consider this case?

danny0405 replied:
I think we can use table name now, can extent it to path if needed.

danny0405 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/SchemaEvolutionContext.java:
```diff
@@ -98,19 +101,63 @@ public SchemaEvolutionContext(InputSplit split, JobConf job) throws IOException
     }
+    this.metaClient = metaClientOption.isPresent() ? metaClientOption.get() : setUpHoodieTableMetaClient();
+    this.internalSchemaOption = getInternalSchemaFromCache();
+    LOG.info("finish init schema evolution for split: {}", split);
```
Do we need this line for debugging? It looks like not very helpful, maybe we can switch to DEBUG devel or just remove it.

danny0405 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/SchemaEvolutionContext.java:
```diff
@@ -98,19 +101,63 @@ public SchemaEvolutionContext(InputSplit split, JobConf job) throws IOException
+        json -> SerDeHelper.fromJson(json)
+    );
+    if (internalSchema != null && internalSchema.isPresent()) {
+      LOG.info("get internal schema from conf for split: {}" + split);
```
Ditto, I think we should remove this line.

muyihao replied:
Will remove it

danny0405 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/SchemaEvolutionContext.java:
```diff
@@ -139,7 +186,10 @@ public void doEvolutionForRealtimeInputFormat(AbstractRealtimeRecordReader realt
     if (internalSchemaOption.isPresent()) {
-      Schema tableAvroSchema = new TableSchemaResolver(metaClient).getTableAvroSchema();
+      Schema tableAvroSchema = getAvroSchemaFromCache();
+      if (tableAvroSchema == null) {
```
When `internalSchemaOption.isPresent()`, the `getAvroSchemaFromCache()` should also exists, this is decided by the cache.

muyihao replied:
Yes, it seems a bit redundant, I will remove it.

danny0405 commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/hive/HoodieCombineHiveInputFormat.java:
```diff
@@ -375,6 +388,39 @@ public InputSplit[] getSplits(JobConf job, int numSplits) throws IOException {
+        for (String path : uniqTablePaths) {
+          HoodieTableMetaClient metaClient = HoodieTableMetaClient.builder().setBasePath(path).setConf(new HadoopStorageConfiguration(job)).build();
+          TableSchemaResolver schemaUtil = new TableSchemaResolver(metaClient);
+          String avroSchema = schemaUtil.getTableAvroSchema().toString();
```
yeah, the cache would be utilized.

danny0405 commented:
The test failures are not related, will merge it.

muyihao commented:
> The test failures are not related, will merge it.

Thanks for helping to land this PR. :)
